tw-tasks-core/src/main/java/com/transferwise/tasks/handler/TaskHandlerRegistry.java

```diff
@@ -18,19 +22,29 @@
   @Autowired
   private ApplicationContext applicationContext;
 
+  @Autowired
+  private TasksProperties tasksProperties;
+
   private volatile List<ITaskHandler> handlers;
 
   @Override
   public ITaskHandler getTaskHandler(IBaseTask task) {
-    List<ITaskHandler> results = getHandlers().stream().filter(h -> h.handles(task)).collect(Collectors.toList());
-
-    if (results.size() > 1) {
-      String resultsSt = StringUtils.join(results.stream().map(p -> p.getClass().getSimpleName()).collect(Collectors.toList()), ",");
-      throw new IllegalStateException("Too many handlers are able to handle task of type '" + task.getType() + "': " + resultsSt);
-    } else if (results.size() == 1) {
-      return results.get(0);
+    // TODO: Should we add caching here for larger applications?
+    // TODO: Or at least advice about keeping the number of handlers low in docs.
+    // TODO: Applications themselves can create hierarchical handlers themselves being able to handle most types of tasks. 
+    ITaskHandler result = null;
+    for (ITaskHandler taskHandler : getHandlers()) {
+      if (taskHandler.handles(task)) {
+        if (result != null) {
+          List<ITaskHandler> results = getHandlers().stream().filter(h -> h.handles(task)).collect(Collectors.toList());
+          String resultsSt = StringUtils.join(results.stream().map(p -> p.getClass().getSimpleName()).collect(Collectors.toList()), ",");
```

Comment:
This one and above can be a single stream ending with `Collectors.joinining(",")`

Reply:
yep, good idea. Will fix.